{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5a2c0b41-6bf3-4009-b3a7-780ab5d1e058",
   "metadata": {},
   "source": [
    "### ~~This work has been deprioritised due to concerns about scope creep~~\n",
    "~~We may wish to revisit this work in a future CNV paper.~~\n",
    "\n",
    "# This work has been reprioritised to generate the CRT breakpoints needed for faceaway reads analysis\n",
    "\n",
    "### The code immediately below was deprioritised so it's a bit outdated and did not directly contribute to Pf8, but I later had to pick it back up again for Pf8's faceaway analysis (further below in section `Collecting breakpoint evidence for each sample as part of Pf8's faceaway analysis`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9e0cbd0-a6fa-4977-80ca-86807f4aafa7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "import igv_notebook\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from IPython.display import Image\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "def plot_cnv_diagnostic(sample, gene):\n",
    "    display(Image(filename = f\"<INSERT PATH HERE>/malariagen-pf8-cnv-calling/04_gcnv_calls_validation/results/{gene}/{sample}/plot.png\"))\n",
    "\n",
    "def igv(sample, locus):\n",
    "    REF = \"<INSERT PATH HERE>/resource-bundle/PlasmoDB-54_Pfalciparum3D7_Genome.fasta\"\n",
    "    GFF = \"<INSERT PATH HERE>/resource-bundle/snpEff/data/Pfalciparum3D7_PlasmoDB_55/PlasmoDB-55_Pfalciparum3D7.gff\"\n",
    "    meta = pd.read_csv(\"../../assets_pf8/01_paths_to_bams.tsv\", sep = \"\\t\", names = [\"SAMPLE\", \"PATH\"])\n",
    "\n",
    "    assert sample in meta.SAMPLE.tolist()\n",
    "\n",
    "    path_to_bam = meta.loc[meta.SAMPLE == sample, \"PATH\"].values[0]\n",
    "    \n",
    "    igv_notebook.init()\n",
    "    \n",
    "    b = igv_notebook.Browser({\n",
    "        \"reference\": {\n",
    "            \"name\": \"Plasmodium falciparum (3D7)\",\n",
    "            \"fastaPath\": REF,\n",
    "            \"indexed\": False\n",
    "        },\n",
    "        \"locus\": locus\n",
    "    })\n",
    "    \n",
    "    b.load_track({\n",
    "        \"name\": \"genes\",\n",
    "        \"type\": \"annotation\",\n",
    "        \"format\": \"gff3\",\n",
    "        \"path\": GFF,\n",
    "        \"displayMode\": \"COLLAPSED\",\n",
    "    })\n",
    "    \n",
    "    b.load_track({\n",
    "        \"path\": path_to_bam,\n",
    "        \"indexPath\": path_to_bam + \".bai\",\n",
    "        \"format\": \"bam\",\n",
    "        \"type\": \"alignment\",\n",
    "        \"displayMode\": \"EXPANDED\",\n",
    "        \"height\": 500,\n",
    "        \"squishedRowHeight\": 5,\n",
    "        \"viewAsPairs\": \"true\",\n",
    "        \"showSoftClips\": \"true\"\n",
    "    })\n",
    "\n",
    "meta = pd.read_csv(\"../../assets_pf8/Pf_8_samples_20241212.txt\", sep = \"\\t\", usecols = [\"Sample\", \"Study\", \"Country\", \"Year\", \"Population\"])\n",
    "meta.Year = meta.Year.fillna(-1).astype(int)\n",
    "\n",
    "pd.set_option(\"display.max.rows\", None)\n",
    "calls = pd.read_csv(\"../../04_gcnv_calls_validation/2024_12_02_pf8_coverage_calls.tsv\", sep = \"\\t\", usecols = [\"SAMPLE\", \"CRT\"]).rename(columns = {\"SAMPLE\": \"Sample\"})\n",
    "calls = (\n",
    "    calls.reset_index(drop = True)\n",
    "    .merge(meta, on = \"Sample\")\n",
    "    .merge(pd.read_csv(\"full_fws_20241125.tsv\", sep = \"\\t\"), on = \"Sample\")\n",
    "    .merge(pd.read_csv(\"pf-haploatlas-PF3D7_0709000_sample_summary.csv\", usecols = [\"Sample\", \"HaploAtlas exclusion reason\", \"ns_changes\"]))\n",
    ")\n",
    "\n",
    "# Show relevant metadata (from pf-haploatlas and pf8 sample metadata) for CRT duplications\n",
    "calls.loc[calls.CRT == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e2e26a0-e8f9-41fd-9aa9-868bc3dad79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 1: Left breakpoint shows lots of soft clips and green reads (face-away reads, which suggests tandem duplication). Right breakpoint shows loads of green lining up nicely, within MED14. These are the large majority of samples, so if they haven't been analysed below, assume they have this breakpoint. \")\n",
    "plot_cnv_diagnostic(\"PA0193-C\", \"CRT\")\n",
    "igv(\"PA0193-C\", \"Pf3D7_07_v3:398,565-399,609\")\n",
    "igv(\"PA0193-C\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9361620-e233-4af0-af98-0217d9e8632e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 1a: Same breakpoints as Figure 1, but is supposedly a triplication. No new evidence from IGV to explain mechanism of triplication though. Burkina Faso. gDNA. Het SNPs appear to be in 1:2 ratio\")\n",
    "plot_cnv_diagnostic(\"PK0019-C\", \"CRT\")\n",
    "igv(\"PK0019-C\", \"Pf3D7_07_v3:398,565-399,609\")\n",
    "igv(\"PK0019-C\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebd20f2b-8ea6-419b-b63b-f1c4c61b8591",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 1b: Also a triplication. Ghana. gDNA. Het SNPs appear to be in 1:2 ratio\")\n",
    "plot_cnv_diagnostic(\"PF0392-C\", \"CRT\")\n",
    "igv(\"PF0392-C\", \"Pf3D7_07_v3:398,565-399,609\")\n",
    "igv(\"PF0392-C\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "277d3373-4f80-41dd-bec5-6b1fe3b3ab1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 1c: Also a triplication. Mali. gDNA. Het SNPs appear to be in 1:2 ratio\")\n",
    "plot_cnv_diagnostic(\"PM0313-C\", \"CRT\")\n",
    "igv(\"PM0313-C\", \"Pf3D7_07_v3:398,565-399,609\")\n",
    "igv(\"PM0313-C\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f8bd1f8-640b-4c4f-8771-59829dae0e35",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 1d: From Laos. This is a strange case though. Potentially DTD? Looks a bit different to Figure 1, but that may be because it's sWGA. The left breakpoint is a little messy to make out, but it helps to use look at the IGV plot for the right breakpoint, scroll down, and look at the mate start values for the faceaway reads there (Look for polyT red). I think it's safest here to assume sWGA is making it appear like a DTD, when really, it's the classic SCO1-MED14 breakpoint. Also, het SNPs are typically in a 50:50 or 55:45 ratio. \")\n",
    "plot_cnv_diagnostic(\"RCN09132\", \"CRT\")\n",
    "igv(\"RCN09132\", \"Pf3D7_07_v3:394,081-405,493\")\n",
    "igv(\"RCN09132\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2467f06-7c62-4410-8323-060cc71a6227",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 2: Different breakpoint to Figure 1. S8e-Cg7. The right breakpoint is further along than MED14, in the middle of Cg7. From Vietnam. \")\n",
    "plot_cnv_diagnostic(\"RCN08130\", \"CRT\")\n",
    "igv(\"RCN08130\", \"Pf3D7_07_v3:364,093-368,572\")\n",
    "igv(\"RCN08130\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b2f6ccb-9ec4-4efa-97c6-7b461967d435",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 2a: Another S8e-Cg7. From Vietnam. Figures 2 and 2a are the only samples here that are homozygous duplications. All others have het duplications\")\n",
    "plot_cnv_diagnostic(\"RCN08145\", \"CRT\")\n",
    "igv(\"RCN08145\", \"Pf3D7_07_v3:364,093-368,572\")\n",
    "igv(\"RCN08145\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c99b759c-4e1c-477a-8452-ed39aecdc2dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 3: COX4-MED14. From Kenya\")\n",
    "plot_cnv_diagnostic(\"SPT42947\", \"CRT\")\n",
    "igv(\"SPT42947\", \"Pf3D7_07_v3:389,424-391,836\")\n",
    "igv(\"SPT42947\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec4d84c8-7c7e-4853-8de1-b49e2a2c728c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Figure 4: HSP110-MED14. From Ghana\")\n",
    "plot_cnv_diagnostic(\"PF0219-C\", \"CRT\")\n",
    "igv(\"PF0219-C\", \"Pf3D7_07_v3:392,203-394,615\")\n",
    "igv(\"PF0219-C\", \"Pf3D7_07_v3:415,952-426,679\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dab5be21-aef7-4520-8a78-362c7f57b87b",
   "metadata": {},
   "source": [
    "---\n",
    "# Making breakpoints file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab48330-d434-41a6-bf15-557ee2a8b769",
   "metadata": {},
   "outputs": [],
   "source": [
    "breakpoints_calls = {\n",
    "    \"SCO1-MED14-triplication\": {\n",
    "        \"Coords\": \"Pf3D7_07_v3:398,900-421,500\",\n",
    "        \"Samples\": [\n",
    "            \"PK0019-C\", \"PF0392-C\", \"PM0313-C\"\n",
    "        ]\n",
    "    },\n",
    "    \"COX4-MED14\": {\n",
    "        \"Coords\": \"Pf3D7_07_v3:390,500-421,500\",\n",
    "        \"Samples\": [\n",
    "            \"SPT42947\"\n",
    "        ]\n",
    "    },\n",
    "    \"HSP110-MED14\": {\n",
    "        \"Coords\": \"Pf3D7_07_v3:390,500-421,500\",\n",
    "        \"Samples\": [\n",
    "            \"PF0219-C\"\n",
    "        ]\n",
    "    },\n",
    "    \"S8e-Cg7\": {\n",
    "        \"Coords\": \"Pf3D7_07_v3:366,000-424,800\",\n",
    "        \"Samples\": [\n",
    "            \"RCN08130\", \"RCN08145\"\n",
    "        ]\n",
    "    },\n",
    "    \"SCO1-MED14\": {\n",
    "        \"Coords\": \"Pf3D7_07_v3:398,900-421,500\",\n",
    "        \"Samples\": \"*\"\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b606d3ba-047c-4a7b-884f-e276a8f8c825",
   "metadata": {},
   "outputs": [],
   "source": [
    "# By looking at various IGV plots, I built up this dictionary which stores information of all the different CRT breakpoints I could find. \n",
    "# Below, we use the dictionary to modify the sample metadata to provide breakpoint information, before doing some summary statistics. \n",
    "\n",
    "crt_breakpoints = calls.copy(deep = True)\n",
    "\n",
    "crt_breakpoints.loc[crt_breakpoints.CRT != 1, \"CRT_breakpoint\"] = \"-\"\n",
    "\n",
    "for k, v in breakpoints_calls.items():\n",
    "    breakpoint_name   = k\n",
    "    breakpoint_coords = v[\"Coords\"]\n",
    "    samples           = v[\"Samples\"]\n",
    "\n",
    "    if samples != \"*\":\n",
    "        crt_breakpoints.loc[crt_breakpoints.Sample.isin(samples), \"CRT_duplication_type\"] = breakpoint_name\n",
    "        crt_breakpoints.loc[crt_breakpoints.Sample.isin(samples), \"CRT_breakpoint\"]       = breakpoint_coords\n",
    "\n",
    "    if samples == \"*\":\n",
    "        crt_breakpoints.loc[crt_breakpoints.CRT == 1, \"CRT_duplication_type\"] = crt_breakpoints.loc[crt_breakpoints.CRT == 1, \"CRT_duplication_type\"].fillna(breakpoint_name)\n",
    "        crt_breakpoints.loc[crt_breakpoints.CRT_duplication_type == breakpoint_name, \"CRT_breakpoint\"] = breakpoint_coords\n",
    "        \n",
    "        break # make sure this step gets performed last, otherwise fillna\n",
    "    \n",
    "crt_breakpoints.CRT = crt_breakpoints.CRT.map({0: \"no_dup\", 1: \"dup\", -1: \"uncallable\"})\n",
    "\n",
    "pd.set_option(\"display.max_colwidth\", None)\n",
    "\n",
    "crt_breakpoints.loc[crt_breakpoints.CRT == \"dup\"].groupby(\"CRT_duplication_type\").apply(lambda s: pd.Series({\n",
    "    \"Breakpoint coordinates\": s.CRT_breakpoint.values[0],\n",
    "    \"Country\": s.Country.value_counts().to_dict(),\n",
    "    \"Year\": s.Year.value_counts().to_dict(),\n",
    "    \"Study\": s.Study.value_counts().to_dict(),\n",
    "}), include_groups=False).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b455f0d8-6b76-437b-91e1-df6ebf5f9335",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_ns_changes(s):\n",
    "    \"\"\"For processing ns_changes from pf-haploatlas data\"\"\"\n",
    "    l_l = [[int(num) for num in re.findall(r'\\d+', l)] for l in s.split(\",\")]\n",
    "\n",
    "    if len(l_l) == 1:\n",
    "        if \"*\" in s:\n",
    "            return {\"het\": l_l[0]}\n",
    "        if \"*\" not in s:\n",
    "            return {\"hom\": l_l[0]}\n",
    "\n",
    "    if len(l_l) == 2:\n",
    "        set1 = set(l_l[0])\n",
    "        set2 = set(l_l[1])\n",
    "\n",
    "        phase1_only = list(set1 - set2)\n",
    "        phase2_only = list(set2 - set1)\n",
    "        both_phases = list(set2 & set1)\n",
    "\n",
    "        return {\n",
    "            \"het\": phase1_only + phase2_only,\n",
    "            \"hom\": both_phases\n",
    "        }\n",
    "\n",
    "dup_analysis = crt_breakpoints.loc[crt_breakpoints.CRT == \"dup\"].reset_index(drop = True).copy(deep = True)\n",
    "dup_analysis[\"int_aa_changes\"] = dup_analysis.ns_changes.apply(process_ns_changes)\n",
    "\n",
    "for i, row in dup_analysis.iterrows():\n",
    "    for colour, l_int in row.int_aa_changes.items():\n",
    "        dup_analysis.loc[dup_analysis.Sample == row.Sample, l_int] = colour\n",
    "\n",
    "aa_columns    = sorted([col for col in dup_analysis.columns if isinstance(col, int)])\n",
    "other_columns = [col for col in dup_analysis.columns if isinstance(col, str)]\n",
    "\n",
    "dup_analysis = dup_analysis[other_columns + aa_columns].sort_values([\"CRT_duplication_type\", \"Population\", \"Country\", 74, 326, 371])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71d0510-884d-4d66-9c16-faa1bde7bf49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Please be aware that while preparing this codebase, we were also battling with some bugs in `call_haplotypes` and pf-haploatlas, so the data is not quite right here yet. We are missing Q271E.\n",
    "\n",
    "fig = make_subplots(\n",
    "    rows=1, cols=5, \n",
    "    subplot_titles=[\"Population\", \"Country\", \"Clonality\", \"CRT breakpoint\", \"CRT haplotype SNP position\"],\n",
    "    shared_yaxes=True,\n",
    "    column_widths=[1, 1, 1, 2, 8],\n",
    "    horizontal_spacing=0.005\n",
    ")\n",
    "\n",
    "population_colours = {\n",
    "    \"SA\": \"#4daf4a\",\n",
    "    \"AF-W\": \"#e31a1c\",\n",
    "    \"AF-C\": \"#fd8d3c\",\n",
    "    \"AF-NE\": \"#bb8129\",\n",
    "    \"AF-E\": \"#fecc5c\",\n",
    "    \"AS-S-E\": \"#dfc0eb\",\n",
    "    \"AS-S-FE\": \"#984ea3\",\n",
    "    \"AS-SE-W\": \"#9ecae1\",\n",
    "    \"AS-SE-E\": \"#3182bd\",\n",
    "    \"OC-NG\": \"#f781bf\",\n",
    "}\n",
    "\n",
    "population_index_map = {pop: i for i, pop in enumerate(dup_analysis.Population.unique())}\n",
    "population_colourscale = [[i / (dup_analysis.Population.nunique()-1), population_colours[pop]] for i, pop in enumerate(dup_analysis.Population.unique())]\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Heatmap(\n",
    "        z=dup_analysis.Population.map(population_index_map),\n",
    "        x=[\"\"] * len(dup_analysis),\n",
    "        y=dup_analysis.Sample.values,\n",
    "        colorscale = population_colourscale,\n",
    "        showscale=False,\n",
    "        text = dup_analysis.Population,\n",
    "        texttemplate=\"%{text}\",\n",
    "        hoverinfo=\"none\",\n",
    "        ygap=1\n",
    "    ),\n",
    "    row=1, col=1\n",
    ")\n",
    "\n",
    "country_index_map = {country: i for i, country in enumerate(dup_analysis.Country.unique())}\n",
    "country_colourscale = [[i / (dup_analysis.Country.nunique()-1), px.colors.qualitative.Bold[i]] for i, country in enumerate(dup_analysis.Country.unique())]\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Heatmap(\n",
    "        z=dup_analysis.Country.map(country_index_map),\n",
    "        x=[\"\"] * len(dup_analysis),\n",
    "        y=dup_analysis.Sample.values,\n",
    "        colorscale = country_colourscale,\n",
    "        showscale=False,\n",
    "        text = dup_analysis.Country,\n",
    "        texttemplate=\"%{text}\",\n",
    "        hoverinfo=\"none\",\n",
    "        ygap=1\n",
    "    ),\n",
    "    row=1, col=2\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "fws_colourscale = [\n",
    "    [1, \"rgba(59, 224, 88, 0.8)\"], [0, \"rgba(216, 238, 220, 0.8)\"]\n",
    "]\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Heatmap(\n",
    "        z=pd.cut(dup_analysis.fws, [0, 0.95, 1], labels = [0, 1]).astype(int),\n",
    "        x=[\"\"] * len(dup_analysis),\n",
    "        y=dup_analysis.Sample.values,\n",
    "        colorscale = \"ice\",\n",
    "        showscale=False,\n",
    "        text = pd.cut(dup_analysis.fws, [0, 0.95, 1], labels = [\"Clonal\", \"Mixed\"]),\n",
    "        texttemplate=\"%{text}\",\n",
    "        hoverinfo=\"none\",\n",
    "        ygap=1\n",
    "    ),\n",
    "    row=1, col=3\n",
    ")\n",
    "\n",
    "\n",
    "dup_type_index_map = {dup_type: i for i, dup_type in enumerate(dup_analysis.CRT_duplication_type.unique())}\n",
    "dup_type_colourscale = [[i / (dup_analysis.CRT_duplication_type.nunique()-1), px.colors.qualitative.Pastel[i]] for i, dup_type in enumerate(dup_analysis.CRT_duplication_type.unique())]\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Heatmap(\n",
    "        z=dup_analysis.CRT_duplication_type.map(dup_type_index_map),\n",
    "        x=[\"\"] * len(dup_analysis),\n",
    "        y=dup_analysis.Sample.values,\n",
    "        colorscale = dup_type_colourscale,\n",
    "        showscale=False,\n",
    "        text = dup_analysis.CRT_duplication_type,\n",
    "        texttemplate=\"%{text}\",\n",
    "        hoverinfo=\"none\",\n",
    "        ygap=1\n",
    "    ),\n",
    "    row=1, col=4\n",
    ")\n",
    "\n",
    "\n",
    "haplotype_colourscale = [\n",
    "    [0, \"#99ccff\"],\n",
    "    [1, \"#ffcccc\"]\n",
    "]\n",
    "\n",
    "fig.add_trace(\n",
    "    go.Heatmap(\n",
    "        z=dup_analysis[aa_columns].map(lambda x: {np.nan: np.nan, \"het\": 1, \"hom\": 2}[x]),\n",
    "        x=[str(col) for col in aa_columns],\n",
    "        y=dup_analysis.Sample.values,\n",
    "        colorscale=haplotype_colourscale,\n",
    "        showscale=False,\n",
    "        text=dup_analysis[aa_columns].map(lambda x: {np.nan: \"\", \"het\": \"Het\", \"hom\": \"Hom\"}[x]).values,\n",
    "        texttemplate=\"%{text}\",\n",
    "        hoverinfo=\"none\",\n",
    "        ygap=1\n",
    "    ),\n",
    "    row=1, col=5\n",
    ")\n",
    "fig.update_xaxes(ticklen=10, tickcolor=\"blue\", tickwidth=2)\n",
    "fig.update_xaxes(showgrid=False, row=1, col=4)\n",
    "\n",
    "fig.update_layout(\n",
    "    height=1000,\n",
    "    width=1500,\n",
    "    template=\"plotly_white\"\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28939902-f5e5-4a67-9a34-e9701abed2d4",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---\n",
    "\n",
    "# Collecting breakpoint evidence for each sample as part of Pf8's faceaway analysis\n",
    "\n",
    "#### The code below here is very scrappy, but I used it to hunt for tandem duplication breakpoints found in `tandem_dup_breakpoints_20250128.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c867f6b-65cc-42fa-829f-35e4d21aa328",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_breakpoints_calls = {\n",
    "    \"SCO1-MED14\": {\n",
    "        \"left_breakpoint\" : \"Pf3D7_07_v3:398,869-398,898\",\n",
    "        \"right_breakpoint\": \"Pf3D7_07_v3:421,765-421,808\",\n",
    "        \"Samples\": [\n",
    "            \"PA0193-C\", \"PF0054-C\", \"PF0118-C\", \"PF0190-C\", \"PF0195-C\", \"PF0231-C\", \"PF0250-C\",\n",
    "            \"PF0356-C\", \"PF0449-C\", \"PF0590-C\", \"PF0729-C\", \"PF0796-C\", \"PF0852-C\", \"PF0905-C\",\n",
    "            \"PF0934-C\", \"PF0960-C\", \"PF0961-C\", \"PF0987-C\", \"PF1103-C\", \"PK0023-C\", \"SPT22565\",\n",
    "            \"PK0063-C\", \"PK0066-C\", \"PK0068-C\", \"PK0075-C\", \"PM0124-C\", \"PM0243-C\", \"PM0336-C\",\n",
    "            \"PM0345-C\", \"PM0455-C\", \"PM0476-C\", \"PM0496-C\", \"PM0580-C\", \"QQ0097-C\", \"QV0096-C\",\n",
    "            \"SPT00802\", \"SPT26595\", \"SPT26638\", \n",
    "        ]\n",
    "    },\n",
    "    \"SCO1-MED14-Laos\": {\n",
    "        \"left_breakpoint\" : \"Pf3D7_07_v3:399,263-399,291\",\n",
    "        \"right_breakpoint\": \"Pf3D7_07_v3:422,103-422,136\",\n",
    "        \"Samples\": [\n",
    "            \"RCN09132\", \"RCN11806\", \"RCN15231\", \"RCN15335\"\n",
    "        ]\n",
    "    },\n",
    "    \"COX4-MED14\": {\n",
    "        \"left_breakpoint\" : \"Pf3D7_07_v3:390,397-390,416\",\n",
    "        \"right_breakpoint\": \"Pf3D7_07_v3:421,765-421,808\",\n",
    "        \"Samples\": [\n",
    "            \"SPT42947\"\n",
    "        ]\n",
    "    },\n",
    "    \"HSP110-MED14\": {\n",
    "        \"left_breakpoint\" : \"Pf3D7_07_v3:393,103-393,133\",\n",
    "        \"right_breakpoint\": \"Pf3D7_07_v3:421,765-421,808\",\n",
    "        \"Samples\": [\n",
    "            \"PF0219-C\"\n",
    "        ]\n",
    "    },\n",
    "    \"S8e-Cg7\": {\n",
    "        \"left_breakpoint\" : \"Pf3D7_07_v3:365,968-365,991\",\n",
    "        \"right_breakpoint\": \"Pf3D7_07_v3:425,216-425,236\",\n",
    "        \"Samples\": [\n",
    "            \"RCN08130\", \"RCN08145\"\n",
    "        ]\n",
    "    },\n",
    "    \"SCO1-MED14-triplication\": {\n",
    "        \"left_breakpoint\" : \"Pf3D7_07_v3:398,869-398,898\",\n",
    "        \"right_breakpoint\": \"Pf3D7_07_v3:421,765-421,808\",\n",
    "        \"Samples\": [\n",
    "            \"PF0392-C\", \"PM0313-C\", \"PK0019-C\"\n",
    "        ]\n",
    "    },\n",
    "    \"no_precise_break_point\": {\n",
    "        \"left_breakpoint\" : \"\",\n",
    "        \"right_breakpoint\": \"\",\n",
    "        \"Samples\": [\n",
    "            \"PF0149-C\", \"PK0069-C\"\n",
    "        ]\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97fca6b4-dd37-4c54-81d5-1ba111674b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_breakpoints = crt_breakpoints.loc[crt_breakpoints.CRT == \"dup\", [\"Sample\", \"Country\", \"Year\", \"Population\", \"fws\", \"CRT_breakpoint\", \"CRT_duplication_type\"]].reset_index(drop = True)\n",
    "\n",
    "for k, v in final_breakpoints_calls.items():\n",
    "    final_breakpoints.loc[final_breakpoints.Sample.isin(v[\"Samples\"]), \"left_breakpoint\"]  = v[\"left_breakpoint\"]\n",
    "    final_breakpoints.loc[final_breakpoints.Sample.isin(v[\"Samples\"]), \"right_breakpoint\"] = v[\"right_breakpoint\"]\n",
    "\n",
    "top_sample = final_breakpoints.loc[pd.isnull(final_breakpoints[[\"left_breakpoint\", \"right_breakpoint\"]]).any(axis = 1), \"Sample\"].values[0]\n",
    "\n",
    "final_breakpoints.loc[pd.isnull(final_breakpoints.left_breakpoint)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0692a8fb-e842-4649-8b9f-ae002875f04a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(top_sample)\n",
    "\n",
    "plot_cnv_diagnostic(top_sample, \"CRT\")\n",
    "igv(top_sample, \"Pf3D7_07_v3:398,724-399,043\")\n",
    "igv(top_sample, \"Pf3D7_07_v3:421,611-421,962\")\n",
    "\n",
    "# igv(top_sample, \"Pf3D7_07_v3:399,263-399,291\")\n",
    "# igv(top_sample, \"Pf3D7_07_v3:422,103-422,136\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b8eecd-50e1-4eb3-9691-ae2f9689a2cb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "final_breakpoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1938324-dfe4-4b2c-9e9e-e8e0b06ae29b",
   "metadata": {},
   "outputs": [],
   "source": [
    "igv(\"PD1033-C\", \"Pf3D7_05_v3:928,340-928,359\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f7460f1-0bae-4b4e-abad-7c4840ee8d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "igv(\"PJ0266-C\", \"Pf3D7_12_v3:939,000-940,000\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
