{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for merging the coverage calls with the faceaway data\n",
    "\n",
    "Note that there is no separate faceaway calls file generated at any stage. This data is available within the final CNV file: `20250130_full_cnv_data_pf8.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<INSERT PATH HERE>/malariagen-pf8-cnv-calling/06_final_calls_and_pf8_figures\n",
      "node-14-10\n",
      "Mon Feb 24 02:32:39 PM GMT 2025\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!hostname\n",
    "!date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inputs\n",
    "todays_date = '20250128'\n",
    "\n",
    "breakpoint_read_counts_fn     = f\"../05_faceaway_data_generation/breakpoint_read_counts_{todays_date}.tsv\"\n",
    "tandem_dup_breakpoints_fn     = f\"../05_faceaway_data_generation/tandem_dup_breakpoints_{todays_date}.txt\"\n",
    "dup_trpinv_dup_breakpoints_fn = f\"../05_faceaway_data_generation/dup_trpinv_dup_breakpoints_{todays_date}.txt\"\n",
    "\n",
    "samples_fn    = '../assets_pf8/Pf_8_samples_20241212.txt'\n",
    "gcnv_calls_fn = '../04_gcnv_calls_validation/2024_12_02_pf8_coverage_calls.tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# df_breakpoint_read_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24409, 150)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PfMDR1_dup_1 faceaways</th>\n",
       "      <th>PfMDR1_dup_1 read pairs</th>\n",
       "      <th>PfMDR1_dup_2 faceaways</th>\n",
       "      <th>PfMDR1_dup_2 read pairs</th>\n",
       "      <th>PfMDR1_dup_3 faceaways</th>\n",
       "      <th>PfMDR1_dup_3 read pairs</th>\n",
       "      <th>PfMDR1_dup_4 faceaways</th>\n",
       "      <th>PfMDR1_dup_4 read pairs</th>\n",
       "      <th>PfMDR1_dup_5 faceaways</th>\n",
       "      <th>PfMDR1_dup_5 read pairs</th>\n",
       "      <th>...</th>\n",
       "      <th>PfGCH1_DTD_1 same directions second</th>\n",
       "      <th>PfGCH1_DTD_1 read pairs second</th>\n",
       "      <th>PfGCH1_DTD_2 same directions first</th>\n",
       "      <th>PfGCH1_DTD_2 read pairs first</th>\n",
       "      <th>PfGCH1_DTD_2 same directions second</th>\n",
       "      <th>PfGCH1_DTD_2 read pairs second</th>\n",
       "      <th>PfGCH1_DTD_3 same directions first</th>\n",
       "      <th>PfGCH1_DTD_3 read pairs first</th>\n",
       "      <th>PfGCH1_DTD_3 same directions second</th>\n",
       "      <th>PfGCH1_DTD_3 read pairs second</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>360</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>172</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>227</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>0</td>\n",
       "      <td>311</td>\n",
       "      <td>0</td>\n",
       "      <td>1334</td>\n",
       "      <td>0</td>\n",
       "      <td>531</td>\n",
       "      <td>0</td>\n",
       "      <td>572</td>\n",
       "      <td>0</td>\n",
       "      <td>572</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>573</td>\n",
       "      <td>0</td>\n",
       "      <td>418</td>\n",
       "      <td>0</td>\n",
       "      <td>647</td>\n",
       "      <td>0</td>\n",
       "      <td>899</td>\n",
       "      <td>0</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0010-CW</th>\n",
       "      <td>0</td>\n",
       "      <td>2128</td>\n",
       "      <td>0</td>\n",
       "      <td>279</td>\n",
       "      <td>0</td>\n",
       "      <td>4792</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>992</td>\n",
       "      <td>0</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>804</td>\n",
       "      <td>14</td>\n",
       "      <td>1216</td>\n",
       "      <td>17</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 150 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           PfMDR1_dup_1 faceaways  PfMDR1_dup_1 read pairs  \\\n",
       "Sample                                                       \n",
       "FP0008-C                        0                      101   \n",
       "FP0009-C                        0                      311   \n",
       "FP0010-CW                       0                     2128   \n",
       "\n",
       "           PfMDR1_dup_2 faceaways  PfMDR1_dup_2 read pairs  \\\n",
       "Sample                                                       \n",
       "FP0008-C                        0                      360   \n",
       "FP0009-C                        0                     1334   \n",
       "FP0010-CW                       0                      279   \n",
       "\n",
       "           PfMDR1_dup_3 faceaways  PfMDR1_dup_3 read pairs  \\\n",
       "Sample                                                       \n",
       "FP0008-C                        0                      168   \n",
       "FP0009-C                        0                      531   \n",
       "FP0010-CW                       0                     4792   \n",
       "\n",
       "           PfMDR1_dup_4 faceaways  PfMDR1_dup_4 read pairs  \\\n",
       "Sample                                                       \n",
       "FP0008-C                        0                      210   \n",
       "FP0009-C                        0                      572   \n",
       "FP0010-CW                       0                      200   \n",
       "\n",
       "           PfMDR1_dup_5 faceaways  PfMDR1_dup_5 read pairs  ...  \\\n",
       "Sample                                                      ...   \n",
       "FP0008-C                        0                      210  ...   \n",
       "FP0009-C                        0                      572  ...   \n",
       "FP0010-CW                       0                      200  ...   \n",
       "\n",
       "           PfGCH1_DTD_1 same directions second  \\\n",
       "Sample                                           \n",
       "FP0008-C                                     0   \n",
       "FP0009-C                                     0   \n",
       "FP0010-CW                                    1   \n",
       "\n",
       "           PfGCH1_DTD_1 read pairs second  PfGCH1_DTD_2 same directions first  \\\n",
       "Sample                                                                          \n",
       "FP0008-C                              172                                   0   \n",
       "FP0009-C                              573                                   0   \n",
       "FP0010-CW                             992                                   0   \n",
       "\n",
       "           PfGCH1_DTD_2 read pairs first  PfGCH1_DTD_2 same directions second  \\\n",
       "Sample                                                                          \n",
       "FP0008-C                             152                                    0   \n",
       "FP0009-C                             418                                    0   \n",
       "FP0010-CW                            455                                    1   \n",
       "\n",
       "           PfGCH1_DTD_2 read pairs second  PfGCH1_DTD_3 same directions first  \\\n",
       "Sample                                                                          \n",
       "FP0008-C                              206                                   0   \n",
       "FP0009-C                              647                                   0   \n",
       "FP0010-CW                             804                                  14   \n",
       "\n",
       "           PfGCH1_DTD_3 read pairs first  PfGCH1_DTD_3 same directions second  \\\n",
       "Sample                                                                          \n",
       "FP0008-C                             227                                    0   \n",
       "FP0009-C                             899                                    0   \n",
       "FP0010-CW                           1216                                   17   \n",
       "\n",
       "           PfGCH1_DTD_3 read pairs second  \n",
       "Sample                                     \n",
       "FP0008-C                               61  \n",
       "FP0009-C                              150  \n",
       "FP0010-CW                             522  \n",
       "\n",
       "[3 rows x 150 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_breakpoint_read_counts = pd.read_csv(breakpoint_read_counts_fn, sep='\\t', index_col=0)\n",
    "\n",
    "print(df_breakpoint_read_counts.shape)\n",
    "\n",
    "df_breakpoint_read_counts.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# df_tandem_dup_breakpoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDR1\n",
      "['PfMDR1_dup_1 read pairs' 'PfMDR1_dup_2 read pairs'\n",
      " 'PfMDR1_dup_3 read pairs' 'PfMDR1_dup_4 read pairs'\n",
      " 'PfMDR1_dup_5 read pairs' 'PfMDR1_dup_6 read pairs'\n",
      " 'PfMDR1_dup_7 read pairs' 'PfMDR1_dup_8 read pairs'\n",
      " 'PfMDR1_dup_9 read pairs' 'PfMDR1_dup_10 read pairs'\n",
      " 'PfMDR1_dup_11 read pairs' 'PfMDR1_dup_12 read pairs'\n",
      " 'PfMDR1_dup_13 read pairs' 'PfMDR1_dup_14 read pairs'\n",
      " 'PfMDR1_dup_15 read pairs' 'PfMDR1_dup_16 read pairs'\n",
      " 'PfMDR1_dup_17 read pairs' 'PfMDR1_dup_18 read pairs'\n",
      " 'PfMDR1_dup_19 read pairs' 'PfMDR1_dup_20 read pairs'\n",
      " 'PfMDR1_dup_21 read pairs' 'PfMDR1_dup_22 read pairs'\n",
      " 'PfMDR1_dup_23 read pairs' 'PfMDR1_dup_24 read pairs'\n",
      " 'PfMDR1_dup_25 read pairs' 'PfMDR1_dup_26 read pairs'\n",
      " 'PfMDR1_dup_27 read pairs' 'PfMDR1_dup_28 read pairs']\n",
      "\n",
      "plasmepsin\n",
      "['PfPlasmepsin_2 read pairs' 'PfPlasmepsin_3 read pairs'\n",
      " 'PfPlasmepsin_1 read pairs']\n",
      "\n",
      "GCH1\n",
      "['PfGCH1_dup_1 read pairs' 'PfGCH1_dup_2 read pairs'\n",
      " 'PfGCH1_dup_3 read pairs' 'PfGCH1_dup_4 read pairs'\n",
      " 'PfGCH1_dup_5 read pairs' 'PfGCH1_dup_6 read pairs'\n",
      " 'PfGCH1_dup_7 read pairs' 'PfGCH1_dup_8 read pairs'\n",
      " 'PfGCH1_dup_9 read pairs' 'PfGCH1_dup_10 read pairs']\n",
      "\n",
      "CRT\n",
      "['CRT_dup_1 read pairs' 'CRT_dup_2 read pairs' 'CRT_dup_3 read pairs'\n",
      " 'CRT_dup_4 read pairs' 'CRT_dup_5 read pairs']\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>first_breakpoint_start</th>\n",
       "      <th>first_breakpoint_end</th>\n",
       "      <th>second_breakpoint_start</th>\n",
       "      <th>second_breakpoint_end</th>\n",
       "      <th>start_gene_id</th>\n",
       "      <th>end_gene_id</th>\n",
       "      <th>breakpoint_name</th>\n",
       "      <th>breakpoint_id</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>use</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pf3D7_05_v3</td>\n",
       "      <td>938329</td>\n",
       "      <td>938357</td>\n",
       "      <td>980012</td>\n",
       "      <td>980040</td>\n",
       "      <td>PF3D7_0522900</td>\n",
       "      <td>PF3D7_0523000</td>\n",
       "      <td>42kb around MDR1</td>\n",
       "      <td>PfMDR1_dup_1</td>\n",
       "      <td>MDR1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pf3D7_05_v3</td>\n",
       "      <td>949514</td>\n",
       "      <td>949527</td>\n",
       "      <td>967253</td>\n",
       "      <td>967266</td>\n",
       "      <td>PF3D7_0522900</td>\n",
       "      <td>PF3D7_0523000</td>\n",
       "      <td>17.7kb around MDR1</td>\n",
       "      <td>PfMDR1_dup_2</td>\n",
       "      <td>MDR1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pf3D7_05_v3</td>\n",
       "      <td>947790</td>\n",
       "      <td>947800</td>\n",
       "      <td>962444</td>\n",
       "      <td>962454</td>\n",
       "      <td>PF3D7_0522900</td>\n",
       "      <td>PF3D7_0523000</td>\n",
       "      <td>15kb around MDR1</td>\n",
       "      <td>PfMDR1_dup_3</td>\n",
       "      <td>MDR1</td>\n",
       "      <td>1</td>\n",
       "      <td>Corresponds to PfMDR1_Nair_8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom  first_breakpoint_start  first_breakpoint_end  \\\n",
       "0  Pf3D7_05_v3                  938329                938357   \n",
       "1  Pf3D7_05_v3                  949514                949527   \n",
       "2  Pf3D7_05_v3                  947790                947800   \n",
       "\n",
       "   second_breakpoint_start  second_breakpoint_end  start_gene_id  \\\n",
       "0                   980012                 980040  PF3D7_0522900   \n",
       "1                   967253                 967266  PF3D7_0522900   \n",
       "2                   962444                 962454  PF3D7_0522900   \n",
       "\n",
       "     end_gene_id     breakpoint_name breakpoint_id target_gene  use  \\\n",
       "0  PF3D7_0523000    42kb around MDR1  PfMDR1_dup_1        MDR1    1   \n",
       "1  PF3D7_0523000  17.7kb around MDR1  PfMDR1_dup_2        MDR1    1   \n",
       "2  PF3D7_0523000    15kb around MDR1  PfMDR1_dup_3        MDR1    1   \n",
       "\n",
       "                           note  \n",
       "0                           NaN  \n",
       "1                           NaN  \n",
       "2  Corresponds to PfMDR1_Nair_8  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tandem_dup_breakpoints = pd.read_csv(tandem_dup_breakpoints_fn, sep='\\t')\n",
    "\n",
    "read_pair_columns = collections.OrderedDict()\n",
    "\n",
    "for genename in df_tandem_dup_breakpoints['target_gene'].unique():\n",
    "    read_pair_columns[genename] = (\n",
    "        df_tandem_dup_breakpoints.loc[\n",
    "            ( df_tandem_dup_breakpoints['target_gene'] == genename )\n",
    "            & ( df_tandem_dup_breakpoints['use'] == 1 )\n",
    "            , 'breakpoint_id'\n",
    "        ] + ' read pairs'\n",
    "    ).values\n",
    "    \n",
    "    print(genename)\n",
    "    print(read_pair_columns[genename])\n",
    "    print()\n",
    "\n",
    "df_tandem_dup_breakpoints.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Loading gCNV HMM results and merging with sample meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24409, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PM2_PM3</th>\n",
       "      <th>HRP2</th>\n",
       "      <th>GCH1</th>\n",
       "      <th>MDR1</th>\n",
       "      <th>HRP3</th>\n",
       "      <th>CRT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0010-CW</th>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           PM2_PM3  HRP2  GCH1  MDR1  HRP3  CRT\n",
       "SAMPLE                                         \n",
       "FP0008-C         0     0    -1     0     0    0\n",
       "FP0009-C         0     0     0     0     0    0\n",
       "FP0010-CW       -1    -1    -1    -1    -1   -1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gcnv_calls = pd.read_csv(gcnv_calls_fn, sep='\\t', index_col=0)\n",
    "\n",
    "print(df_gcnv_calls.shape)\n",
    "\n",
    "df_gcnv_calls.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24409, 22)\n",
      "24409\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Country</th>\n",
       "      <th>Admin level 1</th>\n",
       "      <th>Country latitude</th>\n",
       "      <th>Country longitude</th>\n",
       "      <th>Admin level 1 latitude</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>All samples same case</th>\n",
       "      <th>...</th>\n",
       "      <th>QC pass</th>\n",
       "      <th>Exclusion reason</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Sample was in Pf7</th>\n",
       "      <th>PM2_PM3</th>\n",
       "      <th>HRP2</th>\n",
       "      <th>GCH1</th>\n",
       "      <th>MDR1</th>\n",
       "      <th>HRP3</th>\n",
       "      <th>CRT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0010-CW</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Study     Country   Admin level 1  Country latitude  \\\n",
       "FP0008-C   1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0009-C   1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "\n",
       "           Country longitude  Admin level 1 latitude  Admin level 1 longitude  \\\n",
       "FP0008-C          -10.337093               16.565426                -9.832345   \n",
       "FP0009-C          -10.337093               16.565426                -9.832345   \n",
       "FP0010-CW         -10.337093               16.565426                -9.832345   \n",
       "\n",
       "             Year         ENA All samples same case  ... QC pass  \\\n",
       "FP0008-C   2014.0  ERR1081237              FP0008-C  ...    True   \n",
       "FP0009-C   2014.0  ERR1081238              FP0009-C  ...    True   \n",
       "FP0010-CW  2014.0  ERR2889621             FP0010-CW  ...    True   \n",
       "\n",
       "           Exclusion reason  Sample type Sample was in Pf7 PM2_PM3  HRP2  \\\n",
       "FP0008-C       Analysis_set         gDNA              True       0     0   \n",
       "FP0009-C       Analysis_set         gDNA              True       0     0   \n",
       "FP0010-CW      Analysis_set         sWGA              True      -1    -1   \n",
       "\n",
       "           GCH1  MDR1  HRP3  CRT  \n",
       "FP0008-C     -1     0     0    0  \n",
       "FP0009-C      0     0     0    0  \n",
       "FP0010-CW    -1    -1    -1   -1  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all = (\n",
    "    pd.read_csv(samples_fn, delimiter='\\t', index_col=0, low_memory=False)\n",
    "    .join(df_gcnv_calls, how='inner')\n",
    ")\n",
    "\n",
    "print(df_all.shape)\n",
    "print(np.count_nonzero(df_all['QC pass']))\n",
    "\n",
    "df_all.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "\n",
    "# Generating faceaway calls and also integrating faceaway calls with coverage calls\n",
    "\n",
    "See bottom of notebook for guidance on the logic used to integrate the calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDR1\n",
      "Faceaway only calls:\n",
      "MDR1_faceaway_only\n",
      " 0    17827\n",
      "-1     6289\n",
      " 1      293\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "HMM only calls:\n",
      "MDR1\n",
      " 0    12598\n",
      "-1    11153\n",
      " 1      658\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Integrated calls:\n",
      "MDR1_dup_call\n",
      " 0    19512\n",
      "-1     4195\n",
      " 1      702\n",
      "Name: count, dtype: int64\n",
      "\n",
      " ______________________________\n",
      "plasmepsin\n",
      "Faceaway only calls:\n",
      "plasmepsin_faceaway_only\n",
      " 0    17314\n",
      "-1     5022\n",
      " 1     2073\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "HMM only calls:\n",
      "PM2_PM3\n",
      " 0    13467\n",
      "-1    10454\n",
      " 1      488\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Integrated calls:\n",
      "plasmepsin_dup_call\n",
      " 0    17610\n",
      "-1     4719\n",
      " 1     2080\n",
      "Name: count, dtype: int64\n",
      "\n",
      " ______________________________\n",
      "GCH1\n",
      "Faceaway only calls:\n",
      "GCH1_faceaway_only\n",
      " 0    15886\n",
      "-1     6938\n",
      " 1     1585\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "HMM only calls:\n",
      "GCH1\n",
      "-1    11561\n",
      " 0    11095\n",
      " 1     1753\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Integrated calls:\n",
      "GCH1_dup_call\n",
      " 0    17631\n",
      "-1     4134\n",
      " 1     2644\n",
      "Name: count, dtype: int64\n",
      "\n",
      " ______________________________\n",
      "CRT\n",
      "Faceaway only calls:\n",
      "CRT_faceaway_only\n",
      " 0    21048\n",
      "-1     3282\n",
      " 1       79\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "HMM only calls:\n",
      "CRT\n",
      " 0    13677\n",
      "-1    10681\n",
      " 1       51\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Integrated calls:\n",
      "CRT_dup_call\n",
      " 0    22321\n",
      "-1     2002\n",
      " 1       86\n",
      "Name: count, dtype: int64\n",
      "\n",
      " ______________________________\n"
     ]
    }
   ],
   "source": [
    "min_reads_threshold = 40\n",
    "faceaway_proportion_threshold = 0.025\n",
    "\n",
    "def find_breakpoint_name(position: str, gene: str):\n",
    "    \"\"\"\n",
    "    `read_pair_columns[genename][int(position)]` returns something like \"PfMDR1_dup_8 read pairs\", so we use split to get the first substring. This forms the base_string\n",
    "    \n",
    "    \"\"\"\n",
    "    rename_dictionary = {\n",
    "        \"PfMDR1\"      : \"PfMDR1\",\n",
    "        \"PfPlasmepsin\": \"PfPlasmepsin\",\n",
    "        \"PfGCH1\"      : \"PfGCH1\",\n",
    "        \"CRT\"         : \"PfCRT\",\n",
    "    }\n",
    "    \n",
    "    if np.isnan(position):\n",
    "        return \"\"\n",
    "    \n",
    "    base_string = read_pair_columns[genename][int(position)].split(\" \")[0]\n",
    "\n",
    "    parsed_gene_name = base_string.split(\"_\")[0]\n",
    "\n",
    "    if parsed_gene_name not in rename_dictionary.keys():\n",
    "        raise Exception(f\"{parsed_gene_name} not in `rename_dictionary` of this function. \")\n",
    "\n",
    "    final_gene_name = rename_dictionary[parsed_gene_name]\n",
    "    \n",
    "    breakpoint_id_number_str = base_string.rpartition(\"_\")[2]\n",
    "\n",
    "    if len(read_pair_columns[genename]) >= 10:\n",
    "        breakpoint_id_number_str = breakpoint_id_number_str.zfill(2)\n",
    "\n",
    "    return f\"{final_gene_name}_dup_{breakpoint_id_number_str}\"\n",
    "\n",
    "    \n",
    "def determine_max_proportion_column(x):\n",
    "    if np.all(x==0):\n",
    "        return(np.nan)\n",
    "    else:\n",
    "        return(np.argmax(x))\n",
    "\n",
    "reads                = collections.OrderedDict()\n",
    "faceaway_reads       = collections.OrderedDict()\n",
    "faceaway_proportions = collections.OrderedDict()\n",
    "\n",
    "for genename in read_pair_columns:\n",
    "    print(genename)\n",
    "    \n",
    "    reads[genename] = df_breakpoint_read_counts.loc[:, read_pair_columns[genename]].values # Add read counts data in\n",
    "    reads[genename][reads[genename] == 0] = -1 # If the number of reads is 0, temporarily replace 0 with -1 to prevent ZeroDivisionError later\n",
    "    \n",
    "    faceaway_column_names = np.array([x.replace('read pairs', 'faceaways') for x in read_pair_columns[genename]])\n",
    "    faceaway_reads[genename] = df_breakpoint_read_counts.loc[:, faceaway_column_names].values # Add faceaway read counts data in\n",
    "    faceaway_proportions[genename] = ( faceaway_reads[genename] / reads[genename] )\n",
    "    \n",
    "    faceaway_proportions[genename][faceaway_proportions[genename] <= 0] = 0 # The -1 should cause some proportions to be negative. Set these manually to 0\n",
    "    reads[genename][reads[genename] == -1] = 0 # Change the samples with zero reads from -1 back to 0\n",
    "    \n",
    "    # Calculate maximum faceaway reads proportion for each sample. If no faceaway reads in any of the breakpoints, then set to np.nan\n",
    "    max_proportion_column = np.apply_along_axis(determine_max_proportion_column, 1, faceaway_proportions[genename])\n",
    "\n",
    "    df_all[f\"{genename}_breakpoint\"]     = pd.Series(max_proportion_column).apply(lambda x: find_breakpoint_name(x, genename)).values # Attach breakpoint name (e.g., PfMDR1_dup_08)\n",
    "    df_all[f\"{genename}_max_proportion\"] = np.apply_along_axis(max, 1, faceaway_proportions[genename]) # Find the maximum proportion for the gene for each sample\n",
    "    \n",
    "    breakpoints = np.zeros(faceaway_proportions[genename].shape[0], dtype=object) # Create empty vector of size n, where n = number of samples\n",
    "    breakpoints[~np.isnan(max_proportion_column)] = read_pair_columns[genename][\n",
    "        max_proportion_column[~np.isnan(max_proportion_column)].astype(int)\n",
    "    ] # Add names of breakpoints where there are potential breakpoints\n",
    "\n",
    "    breakpoints[breakpoints==0] = '' # If max proportion of faceaway reads is 0, then replace breakpoint with ''\n",
    "\n",
    "    df_all[f\"{genename}_max_breakpoint\"] = breakpoints # Assign the most probable breakpoint name    \n",
    "    df_all[f\"{genename}_min_reads\"]      = np.apply_along_axis(min, 1, reads[genename])\n",
    "    df_all[f\"{genename}_is_callable\"]    = (df_all[f\"{genename}_min_reads\"] >= min_reads_threshold) # Genes are callable only if there are enough reads\n",
    "\n",
    "    ##############################\n",
    "    # Create faceaway only calls #\n",
    "    ##############################\n",
    "    \n",
    "    df_all[f\"{genename}_faceaway_only\"] = ( df_all[f\"{genename}_max_proportion\"] >= faceaway_proportion_threshold ).astype(int) # Add faceaway evidence-based binary bool\n",
    "    \n",
    "    df_all.loc[ # If even the best possible faceaway evidence for the gene is below the threshold for calling, then call it as -1\n",
    "        ( df_all[f\"{genename}_max_proportion\"] < faceaway_proportion_threshold ) & ( df_all[f\"{genename}_max_proportion\"] > 0.0 )\n",
    "    , f\"{genename}_faceaway_only\"\n",
    "    ] = -1\n",
    "\n",
    "    df_all.loc[ # If the number of reads is too low, then just call as -1\n",
    "        ( df_all[f\"{genename}_min_reads\"] < min_reads_threshold )\n",
    "        , f\"{genename}_faceaway_only\"\n",
    "    ] = -1\n",
    "\n",
    "    # Note: `df_all[f\"{genename.replace('plasmepsin', 'PM2_PM3')}\"]` is the HMM-based call column\n",
    "    # Set the integrated amplification call to the HMM-based call if the HMM-based call is 1 - triggering the condition: \"if either evidence suggests 1, then the final call is 1\"\n",
    "    df_all[f\"{genename}_dup_call\"] = df_all[f\"{genename}_faceaway_only\"] \n",
    "    df_all.loc[\n",
    "        ( df_all[f\"{genename.replace('plasmepsin', 'PM2_PM3')}\"] == 1 )\n",
    "        , f\"{genename}_dup_call\"\n",
    "    ] = 1\n",
    "\n",
    "    # Set the integrated amplification call to the HMM-based call if the reads-based evidence is missing and the HMM-based call is 0 - triggering the condition:\n",
    "    # \"if either evidence is missing, then simply take the call suggested by the non-missing evidence\"\n",
    "    df_all.loc[\n",
    "        ( df_all[f\"{genename.replace('plasmepsin', 'PM2_PM3')}\"] == 0 ) & ( df_all[f\"{genename}_faceaway_only\"] == -1 )\n",
    "        , f\"{genename}_dup_call\"\n",
    "    ] = 0\n",
    "\n",
    "    print(\"Faceaway only calls:\")\n",
    "    print(df_all.loc[df_all['QC pass'], f\"{genename}_faceaway_only\"].value_counts(dropna=False))\n",
    "    print(\"\\n\")\n",
    "\n",
    "    print(\"HMM only calls:\")\n",
    "    print(df_all[f\"{genename.replace('plasmepsin', 'PM2_PM3')}\"].value_counts(dropna=False))\n",
    "    print(\"\\n\")\n",
    "    \n",
    "    print(\"Integrated calls:\")\n",
    "    print(df_all.loc[df_all['QC pass'], f\"{genename}_dup_call\"].value_counts(dropna=False))\n",
    "    print(\"\\n\", \"_\" * 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "\n",
    "# Generate final calls dataframe\n",
    "\n",
    "Note that the following cell only generates the dataframe in memory. You must run the cell below it in order to save the dataframe as a file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4099350/3414927841.py:11: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  .replace({\"nodel\": 0, \"uncallable\": -1, \"del\": 1})\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CRT_uncurated_coverage_only</th>\n",
       "      <th>CRT_curated_coverage_only</th>\n",
       "      <th>CRT_breakpoint</th>\n",
       "      <th>CRT_faceaway_only</th>\n",
       "      <th>CRT_final_amplification_call</th>\n",
       "      <th>GCH1_uncurated_coverage_only</th>\n",
       "      <th>GCH1_curated_coverage_only</th>\n",
       "      <th>GCH1_breakpoint</th>\n",
       "      <th>GCH1_faceaway_only</th>\n",
       "      <th>GCH1_final_amplification_call</th>\n",
       "      <th>MDR1_uncurated_coverage_only</th>\n",
       "      <th>MDR1_curated_coverage_only</th>\n",
       "      <th>MDR1_breakpoint</th>\n",
       "      <th>MDR1_faceaway_only</th>\n",
       "      <th>MDR1_final_amplification_call</th>\n",
       "      <th>PM2_PM3_uncurated_coverage_only</th>\n",
       "      <th>PM2_PM3_curated_coverage_only</th>\n",
       "      <th>PM2_PM3_breakpoint</th>\n",
       "      <th>PM2_PM3_faceaway_only</th>\n",
       "      <th>PM2_PM3_final_amplification_call</th>\n",
       "      <th>HRP2_uncurated_coverage_only</th>\n",
       "      <th>HRP2_breakpoint</th>\n",
       "      <th>HRP2_deletion_type</th>\n",
       "      <th>HRP2_final_deletion_call</th>\n",
       "      <th>HRP3_uncurated_coverage_only</th>\n",
       "      <th>HRP3_breakpoint</th>\n",
       "      <th>HRP3_deletion_type</th>\n",
       "      <th>HRP3_final_deletion_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0010-CW</th>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0011-CW</th>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0012-CW</th>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           CRT_uncurated_coverage_only  CRT_curated_coverage_only  \\\n",
       "FP0008-C                             0                          0   \n",
       "FP0009-C                             0                          0   \n",
       "FP0010-CW                           -1                         -1   \n",
       "FP0011-CW                           -1                         -1   \n",
       "FP0012-CW                           -1                         -1   \n",
       "\n",
       "          CRT_breakpoint  CRT_faceaway_only  CRT_final_amplification_call  \\\n",
       "FP0008-C               -                 -1                             0   \n",
       "FP0009-C               -                  0                             0   \n",
       "FP0010-CW              -                  0                             0   \n",
       "FP0011-CW              -                 -1                            -1   \n",
       "FP0012-CW              -                  0                             0   \n",
       "\n",
       "           GCH1_uncurated_coverage_only  GCH1_curated_coverage_only  \\\n",
       "FP0008-C                             -1                          -1   \n",
       "FP0009-C                              0                           0   \n",
       "FP0010-CW                            -1                          -1   \n",
       "FP0011-CW                            -1                          -1   \n",
       "FP0012-CW                             0                           0   \n",
       "\n",
       "          GCH1_breakpoint  GCH1_faceaway_only  GCH1_final_amplification_call  \\\n",
       "FP0008-C                -                   0                              0   \n",
       "FP0009-C                -                   0                              0   \n",
       "FP0010-CW               -                   0                              0   \n",
       "FP0011-CW               -                   0                              0   \n",
       "FP0012-CW               -                   0                              0   \n",
       "\n",
       "           MDR1_uncurated_coverage_only  MDR1_curated_coverage_only  \\\n",
       "FP0008-C                              0                           0   \n",
       "FP0009-C                              0                           0   \n",
       "FP0010-CW                            -1                          -1   \n",
       "FP0011-CW                            -1                          -1   \n",
       "FP0012-CW                            -1                          -1   \n",
       "\n",
       "          MDR1_breakpoint  MDR1_faceaway_only  MDR1_final_amplification_call  \\\n",
       "FP0008-C                -                   0                              0   \n",
       "FP0009-C                -                   0                              0   \n",
       "FP0010-CW               -                   0                              0   \n",
       "FP0011-CW               -                   0                              0   \n",
       "FP0012-CW               -                   0                              0   \n",
       "\n",
       "           PM2_PM3_uncurated_coverage_only  PM2_PM3_curated_coverage_only  \\\n",
       "FP0008-C                                 0                              0   \n",
       "FP0009-C                                 0                              0   \n",
       "FP0010-CW                               -1                             -1   \n",
       "FP0011-CW                               -1                             -1   \n",
       "FP0012-CW                               -1                             -1   \n",
       "\n",
       "          PM2_PM3_breakpoint  PM2_PM3_faceaway_only  \\\n",
       "FP0008-C                   -                      0   \n",
       "FP0009-C                   -                      0   \n",
       "FP0010-CW                  -                      0   \n",
       "FP0011-CW                  -                      0   \n",
       "FP0012-CW                  -                      0   \n",
       "\n",
       "           PM2_PM3_final_amplification_call  HRP2_uncurated_coverage_only  \\\n",
       "FP0008-C                                  0                             0   \n",
       "FP0009-C                                  0                             0   \n",
       "FP0010-CW                                 0                            -1   \n",
       "FP0011-CW                                 0                            -1   \n",
       "FP0012-CW                                 0                            -1   \n",
       "\n",
       "          HRP2_breakpoint HRP2_deletion_type  HRP2_final_deletion_call  \\\n",
       "FP0008-C                -                  -                         0   \n",
       "FP0009-C                -                  -                         0   \n",
       "FP0010-CW               -                  -                        -1   \n",
       "FP0011-CW               -                  -                        -1   \n",
       "FP0012-CW               -                  -                        -1   \n",
       "\n",
       "           HRP3_uncurated_coverage_only HRP3_breakpoint HRP3_deletion_type  \\\n",
       "FP0008-C                              0               -                  -   \n",
       "FP0009-C                              0               -                  -   \n",
       "FP0010-CW                            -1               -                  -   \n",
       "FP0011-CW                            -1               -                  -   \n",
       "FP0012-CW                             0               -                  -   \n",
       "\n",
       "           HRP3_final_deletion_call  \n",
       "FP0008-C                          0  \n",
       "FP0009-C                          0  \n",
       "FP0010-CW                        -1  \n",
       "FP0011-CW                        -1  \n",
       "FP0012-CW                         0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in uncurated coverage calls for inclusion into the final file for scientific transparency\n",
    "raw_hmm = pd.read_csv(\"../04_gcnv_calls_validation/draft_coverage_calls.tsv\", sep = \"\\t\", index_col = 0)\n",
    "raw_hmm.columns = [f\"{gene}_uncurated_coverage_only\" for gene in raw_hmm.columns]\n",
    "\n",
    "hrp_calls = (\n",
    "    pd.read_csv(\"../04_gcnv_calls_validation/analysis_notebooks/hrp_calls_pf8_20241220.tsv\", sep = \"\\t\")\n",
    "    .rename(columns = {\n",
    "        \"HRP2\": \"HRP2_final_deletion_call\",\n",
    "        \"HRP3\": \"HRP3_final_deletion_call\",\n",
    "    })\n",
    "    .set_index(\"Sample\")\n",
    "    .replace({\"nodel\": 0, \"uncallable\": -1, \"del\": 1})\n",
    "    .fillna(\"-\")\n",
    ")\n",
    "\n",
    "final_cnv_file = (\n",
    "    df_all\n",
    "    .rename(columns = {\n",
    "        \"PM2_PM3\": \"PM2_PM3_curated_coverage_only\",\n",
    "        \"GCH1\":       \"GCH1_curated_coverage_only\",\n",
    "        \"MDR1\":       \"MDR1_curated_coverage_only\",\n",
    "        \"CRT\":         \"CRT_curated_coverage_only\",\n",
    "        \n",
    "        \"HRP3\":       \"HRP3_curated_coverage_only\",\n",
    "        \"HRP2\":       \"HRP2_curated_coverage_only\",\n",
    "        \n",
    "        \"plasmepsin_dup_call\": \"PM2_PM3_final_amplification_call\",\n",
    "        \"HRP2_dup_call\":       \"HRP2_final_amplification_call\",\n",
    "        \"GCH1_dup_call\":       \"GCH1_final_amplification_call\",\n",
    "        \"MDR1_dup_call\":       \"MDR1_final_amplification_call\",\n",
    "        \"HRP3_dup_call\":       \"HRP3_final_amplification_call\",\n",
    "        \"CRT_dup_call\":         \"CRT_final_amplification_call\",\n",
    "\n",
    "        \"plasmepsin_breakpoint\": \"PM2_PM3_breakpoint\",\n",
    "        \"plasmepsin_faceaway_only\": \"PM2_PM3_faceaway_only\"\n",
    "    })\n",
    "    .replace({\"\": \"-\"})\n",
    "    .join(raw_hmm)\n",
    "    .join(hrp_calls)\n",
    "    [[\n",
    "        'CRT_uncurated_coverage_only', 'CRT_curated_coverage_only', 'CRT_breakpoint',\n",
    "        'CRT_faceaway_only', 'CRT_final_amplification_call',\n",
    "        \n",
    "        'GCH1_uncurated_coverage_only', 'GCH1_curated_coverage_only', 'GCH1_breakpoint',\n",
    "        'GCH1_faceaway_only', 'GCH1_final_amplification_call',\n",
    "        \n",
    "        'MDR1_uncurated_coverage_only', 'MDR1_curated_coverage_only', 'MDR1_breakpoint',\n",
    "        'MDR1_faceaway_only', 'MDR1_final_amplification_call',\n",
    "        \n",
    "        'PM2_PM3_uncurated_coverage_only', 'PM2_PM3_curated_coverage_only', 'PM2_PM3_breakpoint',\n",
    "        'PM2_PM3_faceaway_only', 'PM2_PM3_final_amplification_call',\n",
    "        \n",
    "        'HRP2_uncurated_coverage_only', 'HRP2_breakpoint',\n",
    "        'HRP2_deletion_type', 'HRP2_final_deletion_call',\n",
    "        \n",
    "        'HRP3_uncurated_coverage_only', 'HRP3_breakpoint',\n",
    "        'HRP3_deletion_type', 'HRP3_final_deletion_call',\n",
    "    ]]\n",
    ")\n",
    "\n",
    "for cnv_gene in [\"CRT\", \"GCH1\", \"MDR1\", \"PM2_PM3\"]:\n",
    "    final_cnv_file.loc[final_cnv_file[f\"{cnv_gene}_faceaway_only\"] != 1, f\"{cnv_gene}_breakpoint\"] = \"-\"\n",
    "\n",
    "pd.set_option(\"display.max.columns\", None)\n",
    "\n",
    "final_cnv_file.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Exporting final CNV file\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exporting 20250130_full_cnv_data_pf8.tsv\n",
      "20250130_full_cnv_data_pf8.tsv exported.\n"
     ]
    }
   ],
   "source": [
    "export_file_name = \"20250130_full_cnv_data_pf8.tsv\"\n",
    "print(\"Exporting\", export_file_name)\n",
    "\n",
    "final_cnv_file.reset_index().rename(columns = {\"index\": \"Sample\"}).to_csv(export_file_name, sep = \"\\t\", index = False)\n",
    "\n",
    "print(export_file_name, \"exported.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Appendix: The way amplifications and deletions are handled was different - please read carefully\n",
    "\n",
    "---\n",
    "\n",
    "### For amplifications\n",
    "\n",
    "Columns containing values, `{-1, 0, 1}`, are genotype columns:\n",
    "\n",
    "- `-1` indicates missing evidence\n",
    "- `0` indicates evidence for **_no_** amplification\n",
    "- `1` indicates evidence for amplification\n",
    "\n",
    "Genotype columns ending in `_only` are intermediate forms of evidence, which are integrated to create columns ending in `_final_amplification_call`:\n",
    "\n",
    "- `_uncurated_coverage_only` are raw genotypes generated from the depth/coverage-based gCNV HMM pipeline\n",
    "- `_curated_coverage_only` are modified genotypes based on `_uncurated_coverage_only` that were manually curated for the lowest confidence samples\n",
    "- `_faceaway_only` are the genotypes from Richard's reads-based faceaway detection pipeline\n",
    "\n",
    "Also, the `_breakpoint` column contains the names of breakpoints (e.g., `CRT_dup_3`) that were found from the process that generated the `_faceaway_only` data.\n",
    "\n",
    "The `_final_amplification_call` genotypes were generated using only the `_curated_coverage_only` and `_faceaway_only` genotypes. These two sources of evidence had equal priority and were integrated using the following rules:\n",
    "\n",
    "1. If any of the genotypes are `1`, the final call is `1`\n",
    "2. If both genotypes are `-1` (missing), the final call is `-1`\n",
    "3. If only one of the genotypes are `-1` (missing), the final call inherits the value of the non-missing source of evidence\n",
    "\n",
    "---\n",
    "\n",
    "### For deletions\n",
    "\n",
    "Columns containing values, `{-1, 0, 1}`, are genotype columns:\n",
    "\n",
    "- `-1` indicates missing evidence\n",
    "- `0` indicates evidence for **_no_** amplification\n",
    "- `1` indicates evidence for amplification\n",
    "\n",
    "Genotype columns ending in `_only` are intermediate forms of evidence, which are integrated to create columns ending in `_final_amplification_call`:\n",
    "\n",
    "- `_uncurated_coverage_only` are raw genotypes generated from the depth/coverage-based gCNV HMM pipeline\n",
    "\n",
    "The `_final_deletion_call` genotypes were generated by manually curating the `_uncurated_coverage_only` genotypes. The manual curation stage for gene deletions required for there to be breakpoints visible (on IGV) for them to be accepted as deletions in the `_final_deletion_call` genotypes (otherwise they are called as `-1`) and hence, there is no `curated_coverage_only` genotype as that is essentially the `_final_deletion_call` already. Richard's reads-based faceaway detection pipeline was **_not_** used in calling deletions. \n",
    "\n",
    "During manual curation, the `_breakpoint` and `_deletion_type` columns were also populated. \n",
    "\n",
    "---\n",
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
